modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -374,40 +366,47 @@ private void saveGroupConfiguration(AbstractWriteBatch writeBatch, byte[] config
 
             assert rowIsLocked(rowId);
 
-            ByteBuffer keyBuf = prepareHeapKeyBuf(rowId);
+            ByteBuffer keyBuf = prepareDirectKeyBuf(rowId)
+                    .position(0)
+                    .limit(ROW_PREFIX_SIZE);
 
             BinaryRow res = null;
 
             try {
                 // Check concurrent transaction data.
-                byte[] keyBufArray = keyBuf.array();
+                byte[] keyBytes = new byte[ROW_PREFIX_SIZE];
 
-                byte[] keyBytes = copyOf(keyBufArray, ROW_PREFIX_SIZE);
+                keyBuf.get(keyBytes);
 
-                byte[] previousValue = writeBatch.getFromBatchAndDB(db, helper.partCf, readOpts, keyBytes);
+                keyBuf.rewind();
+
+                byte[] previousValueBytes = writeBatch.getFromBatchAndDB(db, helper.partCf, readOpts, keyBytes);
 
                 // Previous value must belong to the same transaction.
-                if (previousValue != null) {
+                if (previousValueBytes != null) {
+                    ByteBuffer previousValue = ByteBuffer.wrap(previousValueBytes);
+
                     validateTxId(previousValue, txId);
 
                     res = wrapValueIntoBinaryRow(previousValue, true);
                 }
 
                 if (row == null) {
+                    ByteBuffer value = allocateDirect(VALUE_HEADER_SIZE);
```

Comment:
I would avoid frequent usage of allocateDirect, it will create unnecessary GC pressure, as far as I know. That's why in the old code all direct buffers are cached as thread-locals.

Reply:
> it will create unnecessary GC pressure

Can you please clarify? What's the difference between using a direct buffer or copying a byte array inside JNI?

Reply:
ok, I did some research, looks like this is indeed a bad idea. I'll rollback this change